finality-aleph/src/new_network/substrate.rs

```diff
@@ -21,19 +22,30 @@ impl<B: Block, H: ExHashT> RequestBlocks<B> for Arc<NetworkService<B, H>> {
     }
 }
 
+#[async_trait]
 impl<B: Block, H: ExHashT> Network for Arc<NetworkService<B, H>> {
     fn event_stream(&self) -> NetworkEventStream {
         Box::pin(self.as_ref().event_stream("aleph-network"))
     }
 
-    fn message_sender(
-        &self,
+    async fn send<'a>(
+        &'a self,
+        data: impl Into<Vec<u8>> + Send + Sync + 'static,
         peer_id: PeerId,
         protocol: Cow<'static, str>,
-    ) -> Option<NotificationSender> {
-        self.notification_sender(peer_id.into(), protocol)
-            .map_err(|_| debug!(target: "aleph-network", "Attempted send to peer we are not connected to."))
-            .ok()
+    ) -> Result<(), NetworkSendError> {
+        if let Ok(sender) = self.notification_sender(peer_id.into(), protocol) {
```

Comment:
It's probably better to `map_err` this and `?` after that. Just make the error here more expressive (see other comments about that). This whole method can probably be remade into 3 calls in succession which do `map_err` into `?`. It could even be a "one liner" like `self.notification_sender(peer_id.into(), protocol).map_err(...)?.ready().await.map_err(...)?.send(data).map_err(...);`. All the logging should happen outside anyway, only the returned error should contain enough information to be useful.

Reply:
Done